`key-spacing`: `multiLine.align.mode` is overridden, `multiLine.mode` with `align` causes an error

**What version of ESLint are you using?**
ESLint 3.1.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

case1.js
```javascript
module.exports = {
    "rules"   : {
        "key-spacing"                  : [
            "error",
            {
                "multiLine" : {
                    "beforeColon": false,
                    "afterColon" : true,
                    "mode"       : "strict",
                    "align"      : {
                        "beforeColon": false,
                        "afterColon" : true,
                        "on"         : "colon",
                        "mode"       : "minimum"
                    }
                }
            }
        ]
    }
};
```

case2.js
```javascript
module.exports = {
    "rules": {
        "key-spacing": [
            "error",
            {
                "multiLine": {
                    "beforeColon": false,
                    "afterColon" : true,
                    "mode"       : "strict"
                },
                "align"    : {
                    "beforeColon": false,
                    "afterColon" : true,
                    "on"         : "colon",
                    "mode"       : "minimum"
                }
            }
        ]
    }
};
```

**What did you do? Please include the actual source code causing the issue.**

test.js
```javascript
var a = {
    foo   : 1,
    align1: {
        foo:    2
    },
    align2: {
        foo:    3
    },
    bar: 4
};
```

**What did you expect to happen?**

Case 1
```shell
  5:5   warning  Unexpected a line break before this close brace  object-curly-newline
  6:13  warning  Unexpected a line break after this open brace    object-curly-newline
```

Case2
```shell
  1:1   error    Unexpected var, use let or const instead         no-var
  1:5   error    'a' is defined but never used                    no-unused-vars
  3:13  warning  Unexpected a line break after this open brace    object-curly-newline
  5:5   warning  Unexpected a line break before this close brace  object-curly-newline
  6:13  warning  Unexpected a line break after this open brace    object-curly-newline
  8:5   warning  Unexpected a line break before this close brace  object-curly-newline
```

**What actually happened? Please include the actual, raw output from ESLint.**

Case 1
```shell
  4:17  error    Extra space before value for key 'foo'           key-spacing
  5:5   warning  Unexpected a line break before this close brace  object-curly-newline
  6:13  warning  Unexpected a line break after this open brace    object-curly-newline
  7:17  error    Extra space before value for key 'foo'           key-spacing
```
Because of https://github.com/eslint/eslint/blob/67c3cc2fa655e9e92e01d7a88e9a9d60cd4a4ce9/lib/rules/key-spacing.js#L139

Case2
```shell
/path/to/config/case2.js:
        Configuration for rule "key-spacing" is invalid:
        Value "[object Object]" no schemas match.

Error: /path/to/config/case2.js:
        Configuration for rule "key-spacing" is invalid:
        Value "[object Object]" no schemas match.
```

Because of https://github.com/eslint/eslint/blob/67c3cc2fa655e9e92e01d7a88e9a9d60cd4a4ce9/lib/rules/key-spacing.js#L271-L321 (`multiLine.mode` isn't allowed with `align` option)

Are they designed as they are?

@btmills Thank you for you explanation.  I had a slight misunderstanding about *align*... :flushed: 
I tried to lint the first code in [align and multiLine](https://github.com/eslint/eslint/blob/67c3cc2fa655e9e92e01d7a88e9a9d60cd4a4ce9/docs/rules/key-spacing.md#align-and-multiline) with `case1.js` config and the code actually passed, so I also tried to lint another code based on the code above.

```javascript
var myObj = {
  key1: 1, // uses multiLine

  key2:    2, // uses align (when defined)
  key3:    3, // uses align (when defined)

  key4: 4 // uses multiLine
}
```

The result is:
```shell
  4:12  error  Extra space before value for key 'key2'  key-spacing
  5:12  error  Extra space before value for key 'key3'  key-spacing
```
I'm a little confused because `multiLine.align.mode`  in `case1.js` seems to be overridden with `multiLine.mode`. Is this also a bug? Or we cannot use both `multiLine.mode` and `multiLine.align.mode` in the same time?

--- a/lib/rules/key-spacing.js
+++ b/lib/rules/key-spacing.js
@@ -136,7 +136,7 @@ function initOptions(toOptions, fromOptions) {
         if (toOptions.multiLine.align) {
             toOptions.align = {
                 on: toOptions.multiLine.align.on,
-                mode: toOptions.multiLine.mode,
+                mode: toOptions.multiLine.align.mode || toOptions.multiLine.mode,
                 beforeColon: toOptions.multiLine.align.beforeColon,
                 afterColon: toOptions.multiLine.align.afterColon
             };
@@ -289,6 +289,9 @@ module.exports = {
                         multiLine: {
                             type: "object",
                             properties: {
+                                mode: {
+                                    enum: ["strict", "minimum"]
+                                },
                                 beforeColon: {
                                     type: "boolean"
                                 },

--- a/tests/lib/rules/key-spacing.js
+++ b/tests/lib/rules/key-spacing.js
@@ -711,6 +711,56 @@ ruleTester.run("key-spacing", rule, {
             }
         }],
         parserOptions: { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } }
+    }, {
+        code: [
+            "var obj = {",
+            "    key1: 1,",
+            "",
+            "    key2:    2,",
+            "    key3:    3,",
+            "",
+            "    key4: 4",
+            "}"
+        ].join("\n"),
+        options: [{
+            multiLine: {
+                beforeColon: false,
+                afterColon: true,
+                mode: "strict",
+                align: {
+                    beforeColon: false,
+                    afterColon: true,
+                    on: "colon",
+                    mode: "minimum"
+                }
+            }
+        }],
+        parserOptions: { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } }
+    }, {
+        code: [
+            "var obj = {",
+            "    key1: 1,",
+            "",
+            "    key2:    2,",
+            "    key3:    3,",
+            "",
+            "    key4: 4",
+            "}"
+        ].join("\n"),
+        options: [{
+            multiLine: {
+                beforeColon: false,
+                afterColon: true,
+                mode: "strict"
+            },
+            align: {
+                beforeColon: false,
+                afterColon: true,
+                on: "colon",
+                mode: "minimum"
+            }
+        }],
+        parserOptions: { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } }
     }],
 
     invalid: [{
@@ -1621,6 +1671,62 @@ ruleTester.run("key-spacing", rule, {
             { message: "Extra space before value for key \'xs\'.", line: 7, column: 21, type: "Literal" },
             { message: "Extra space before value for key \'func2\'.", line: 8, column: 16, type: "FunctionExpression" },
             { message: "Extra space after key \'singleLine\'.", line: 11, column: 5, type: "Identifier" }
+        ]
+    }, {
+        code: [
+            "var obj = {",
+            "    key1: 1,",
+            "",
+            "    key2:    2,",
+            "    key3:    3,",
+            "",
+            "    key4: 4",
+            "}"
+        ].join("\n"),
+        options: [{
+            multiLine: {
+                beforeColon: false,
+                afterColon: true,
+                mode: "strict",
+                align: {
+                    beforeColon: false,
+                    afterColon: true,
+                    on: "colon",
+                }
+            }
+        }],
+        parserOptions: { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } },
+        errors: [
+            { message: "Extra space before value for key \'key2\'.", line: 4, column: 14, type: "Literal" },
+            { message: "Extra space before value for key \'key3\'.", line: 5, column: 14, type: "Literal" }
+        ]
+    }, {
+        code: [
+            "var obj = {",
+            "    key1: 1,",
+            "",
+            "    key2:    2,",
+            "    key3:    3,",
+            "",
+            "    key4: 4",
+            "}"
+        ].join("\n"),
+        options: [{
+            multiLine: {
+                beforeColon: false,
+                afterColon: true,
+                mode: "strict"
+            },
+            align: {
+                beforeColon: false,
+                afterColon: true,
+                on: "colon",
+            }
+        }],
+        parserOptions: { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } },
+        errors: [
+            { message: "Extra space before value for key \'key2\'.", line: 4, column: 14, type: "Literal" },
+            { message: "Extra space before value for key \'key3\'.", line: 5, column: 14, type: "Literal" }
         ]
     }]
 });

